ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -56,81 +54,125 @@ import           Ouroboros.Network.Mux.Types
-    -> Peer ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Pipelined client and a server with the correspnding codec.
-  --
-  PipelinedClientAndServer
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
+-- * enumeration of server applications: this application type is mostly useful
+--   tests.
+--
+-- * enumeration of both client and server applications, e.g. a full node
+--   serving downstream peers using server side of each protocol and getting
+--   updates from upstream peers using client side of each of the protocols.
+--
+data MuxApplication (appType :: AppType) ptcl m  where
+  MuxInitiatorApplication
+    -- Initiator application; most simple application will be @'runPeer'@ or
+    -- @'runPipelinedPeer'@ supplied with a codec and a @'Peer'@ for each
+    -- @ptcl@.  But it allows to handle resources if just application of
+    -- @'runPeer'@ is not enough.  It will be run as @'ModeInitiator'@.
+    :: (ptcl -> Channel m ByteString ->  m ())
+    -> MuxApplication InitiatorApp ptcl m
+
+  MuxResponderApplication
+    -- Responder application; similarly to the @'MuxInitiatorApplication'@ but it
+    -- will be run using @'ModeResponder'@.
+    :: (ptcl -> Channel m ByteString ->  m ())
+    -> MuxApplication ResponderApp ptcl m
+
+  MuxInitiatorAndResponderApplication
+    -- Initiator and server applications.
+    :: (ptcl -> Channel m ByteString ->  m ())
+    -> (ptcl -> Channel m ByteString ->  m ())
+    -> MuxApplication InitiatorAndResponderApp ptcl m
 
+-- |
+-- Accessor for the client side of a @'MuxApplication'@.
+--
+initiatorApplication
+  :: HasInitiator appType ~ True
+  => (MuxApplication appType ptcl m)
+  -> (ptcl -> Channel m ByteString ->  m ())
+initiatorApplication (MuxInitiatorApplication app) = \ptcl channel -> app ptcl channel
+initiatorApplication (MuxInitiatorAndResponderApplication app _) = \ptcl channel -> app ptcl channel
 
 -- |
--- Public network interface for 'ouroboros-network'.
+-- Accessor for the client side of a @'MuxApplication'@.
 --
-data NetworkInterface ptcl addr m r = NetworkInterface {
-      -- |
-      -- Address of the node to run.  The node will bind to this address, and
-      -- listen for incoming connections.  Some bearers do not have a notion of
-      -- address.
-      --
-      nodeAddress      :: addr,
+responderApplication
+  :: HasResponder appType ~ True
+  => (MuxApplication appType ptcl m)
+  -> (ptcl -> Channel m ByteString ->  m ())
+responderApplication (MuxResponderApplication app) = app
+responderApplication (MuxInitiatorAndResponderApplication _ app) = app
+
+-- |
+-- This type is only necessary to use the @'simpleMuxClient'@ and
+-- @'simpleMuxServer'@ smart constructors.
+data MuxPeer failure m a where
```

Comment:
That's true, any suggestions? The `Mux` here is only used because this is used only in the context of `muxStart` function.  With your presentation of session types we would not need this sum at all.

Reply:
The function you are asking for are called: `simpleMuxInitiatorApplication` and `simpleMuxResponderApplication`. There is no need to have `runMuxPeer`.  I will refactor them to get the `runMuxPeer`.